### DISCUSSION TITLE
Some subscriptions not completed by server

### DISCUSSION THREAD

**Discussion by Amareis** [enisdenjo/graphql-ws#402] [Q&A] [answered]
**Screenshot**
![image](https://user-images.githubusercontent.com/4806617/190646185-4d295c60-dfe2-4c3e-b757-6a1ee9e94646.png)

**Expected Behaviour**
I expected that all cancelled subscriptions should be acked with `complete` message from server.

**Actual Behaviour**
but instead two subscriptions is not acked with `complete` - `userAddedToChat 65` and `meAddedToChat 66`, namely.

**Debug Information**
Help us debug the bug?

**Further Information**
Of course, it's a some bug in app server code, some stuck subscriptions or so. But in that case graphql-ws should report it somehow - may be add some timeout on subscriptions completing?

**Comment by Amareis** [comment]
I think if there is stuck subscriptions on server, this must be reported - it's pretty good opportunity for memory leak.

**Comment by enisdenjo** [comment, marked as answer]
The server has exhaustive tests making sure the subscriptions get completed when necessary. I wouldn't say graphql-ws is causing this. Please make sure your subscription generators support the `return` method.

Furthermore, the server does not acknowledge a complete message with another complete message. If the client sent a complete message, the server is expected to complete the subscription - it does not need to notify the client. The server will only send complete messages if the subscription was ended on its side without the client requesting it.

**Reply by Amareis** [reply]
Uhoh, so why it then reports rest of subscriptions?

**Reply by enisdenjo** [reply]
Probably because the subscription generators don't support the `return` method, or are cleaning up late.

See here:
https://github.com/enisdenjo/graphql-ws/blob/5a56344795f3d2524397b6205442f6ac67e9d635/src/server.ts#L823-L825

and here:
https://github.com/enisdenjo/graphql-ws/blob/5a56344795f3d2524397b6205442f6ac67e9d635/src/server.ts#L832-L838

**Reply by Amareis** [reply]
Maybe then graphql-ws should remove subscription before calling return? Currently it's some strange inconsistence behavior (and additional messages on the wires).

**Reply by enisdenjo** [reply]
> Maybe then graphql-ws should remove subscription before calling return?

Yeah, just thinking about that. 🤔

Let me write up a few tests and I'll see what I can do.

**Reply by Amareis** [reply]
Additionally, protocol says that after complete subscriptions no additional messages should be sent about it - and this messages clearly sent after comleting subscriptions.

**Reply by enisdenjo** [reply]
You're absolutely right - I'll take care of it.

Thanks for reporting!

**Reply by Amareis** [reply]
Maybe reopen issue then?

**Reply by enisdenjo** [reply]
Can't original issue is locked. 😬

Nevertheless, what you've discovered is a bit different - I've made a new issue #403.
